// @mockoon/commons/dist/cjs/libs/utils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.RemoveLeadingSlash = exports.IsEqual = exports.CloneObject = exports.UnwrapLegacyExport = exports.IsLegacyExportData = exports.IsValidURL = exports.GetRouteResponseContentType = exports.GetContentType = void 0;
/**
 * Extract the content-type from an array of headers
 *
 * @param headers
 */
const GetContentType = (headers) => {
    const contentTypeHeader = headers.find((header) => header.key.toLowerCase() === 'content-type');
    if (contentTypeHeader) {
        return contentTypeHeader.value;
    }
    return null;
};
exports.GetContentType = GetContentType;
/**
 * Return a route response's content-type.
 * Environment's content-type is overridden by route's content-type
 *
 * @param environment
 * @param routeResponse
 */
const GetRouteResponseContentType = (environment, routeResponse) => {
    const routeResponseContentType = (0, exports.GetContentType)(routeResponse.headers);
    const environmentContentType = (0, exports.GetContentType)(environment.headers);
    return routeResponseContentType || environmentContentType || '';
};
exports.GetRouteResponseContentType = GetRouteResponseContentType;
/**
 * Test if URL is valid
 *
 * @param URL
 */
const IsValidURL = (address) => {
    try {
        const myURL = new URL(address);
        return true;
    }
    catch (e) {
        return false;
    }
};
exports.IsValidURL = IsValidURL;
/**
 * Verify if the object is a Mockoon legacy export data object
 *
 * @param data
 * @returns
 */
const IsLegacyExportData = (data) => {
    var _a;
    return data.source !== undefined &&
        ((_a = data.source) === null || _a === void 0 ? void 0 : _a.split(':')[0]) === 'mockoon';
};
exports.IsLegacyExportData = IsLegacyExportData;
/**
 * Import legacy export Mockoon's format.
 * Data was wrapped and could enclose multiple environments (and routes):
 *
 * ```
 * {
 *   "source": "mockoon:1.17.0",
 *   "data": [
 *     {
 *       "type": "environment",
 *       "item": {
 *         "uuid": "",
 *         "lastMigration": 13,
 *         "name": "Tutorial - Generate mock data"
 *         ...
 *       }
 *     }
 *   ]
 * }
 *
 * ```
 */
const UnwrapLegacyExport = (exportData) => exportData.data.reduce((environments, dataItem) => {
    if (dataItem.type === 'environment') {
        environments.push(dataItem.item);
    }
    return environments;
}, []);
exports.UnwrapLegacyExport = UnwrapLegacyExport;
/**
 * Clone an object using JSON.stringify
 * /!\ Suitable for Environment, Route, etc but not for complex objects containing Map, Set, etc
 */
const CloneObject = (objectToClone) => JSON.parse(JSON.stringify(objectToClone));
exports.CloneObject = CloneObject;
/**
 * Compare two objects using JSON.Stringify
 */
const IsEqual = (firstObject, secondObject) => JSON.stringify(firstObject) === JSON.stringify(secondObject);
exports.IsEqual = IsEqual;
const RemoveLeadingSlash = (str) => str.replace(/^\//g, '');
exports.RemoveLeadingSlash = RemoveLeadingSlash;
